torch/csrc/autograd/engine.cpp

```diff
@@ -73,14 +74,33 @@ auto Engine::backward(const variable_list& variables,
         did_leaf_backward = true;
       }
     } else {
-      creators.push_back(var->creator.get());
-      if (var->creator->requires_grad) {
-        GradBuffer buf(var->creator->num_outputs);
-        buf.addGrad(var->output_nr, Variable::of(std::move(grad)));
-        ready.emplace_front(var->creator, std::move(buf));
+      auto& creator = var->creator;
+      auto it = creator_grad.find(creator);
+      if (it == creator_grad.end()) {
+          std::unique_ptr<GradBuffer> buf;
+        if (creator->requires_grad) {
+          buf.reset(new GradBuffer(creator->num_outputs));
+          buf->addGrad(var->output_nr, Variable::of(std::move(grad)));
+        }
+        creator_grad.emplace(creator, std::move(buf));
+      // Duplicate creator
+      } else {
+        if (creator->requires_grad) {
+          auto& buf = it->second;
+          buf->addGrad(var->output_nr, Variable::of(std::move(grad)));
+        }
       }
     }
   }
+  for (auto& entry: creator_grad) {
+    const auto& creator = entry.first;
+    auto& buf = entry.second; // WARNING: this is nullptr if !creator->requires_grad
```

Comment:
It has entries for those which don't require grad, but the buffer is `nullptr`. [This line](https://github.com/pytorch/pytorch/pull/886/files#diff-01f226b610655c49b1e52eddf9ea4126R85) is not guarded by a `requires_grad` check.